### DISCUSSION TITLE
Pixi VSCode extension

### DISCUSSION THREAD

**Discussion by pavelzw** [prefix-dev/pixi#415] [Ideas]
I think a Pixi VSCode extension that automatically sets your python interpreter if python is in the environment (and maybe also has some other features) would be awesome.

**Comment by pavelzw** [comment]
There is the possibility to choose `<path-to-project>/.pixi/env/bin/python3.11` as python interpreter but VSCode automatically recognizes it as a conda python and tries to run `conda activate <path-to-project>/slim-trees/.pixi/env` on startup.

**Comment by ruben-arts** [comment]
@pavelzw that would be really nice indeed! What other features would that include? We don't have a lot of experience with write plugins for IDE's so any input would be welcome!

**Reply by camstuart** [reply]
From a JetBrains point of view, there are a few features that would be amazing: 
- Create project (existing capability) allows you to select the type of project, EG FastAPI Python. Add the choice 'pixi' to the existing options of conda, poetry, venv etc...
- Set up run configurations. Currently JetBrains understands various task and code runners such as npm scripts, Makefiles and sets up green 'play' buttons that allow execution or debug execution from the entry points. Pixi run commands would work the same way.
- In PyCharm, you can search pypi for dependancies, if poetry is the current dep tool, then it updates pyproject.toml. Instead this could update the pixi dependancy definitions.

One great thing about creating a JetBrains extension is that it could then be used by the various IDE's PyCharm, WebStorm, RustRover and so on. So most of the general pixi capability once created, is then available amongst all these IDE's

**Comment by pavelzw** [comment]
Maybe some tooltips in the `pixi.toml`?

I could also imagine people wanting to execute pixi commands using this thing (I don't know what it's called 😅)

> ![image](https://github.com/prefix-dev/pixi/assets/29506042/041dbb10-b328-410d-bc67-411c27e03e97)

Maybe also on the side bar an overview of installed packages

**Comment by pavelzw** [comment]
Also having something like this button in the `pixi.toml` would be great

![image](https://github.com/prefix-dev/pixi/assets/29506042/4eeaea49-0b33-45cc-8288-b34d56700ccc)

Also, you can hover in vscode over dependencies in the package.json to get some package metadata (and the newest version). I think that would also be nice for the pixi.toml

**Comment by pavelzw** [comment]
Another issue I stumbled on while using pixi with VSCode is that VSCode tries to automatically activate the conda environment.

```bash
❯ source activate
source: no such file or directory: activate
❯ conda activate <path-to-project>/.pixi/env
zsh: command not found: conda
```

The solution is to uncheck this box.
![image](https://github.com/prefix-dev/pixi/assets/29506042/2321b893-4439-4605-8f2f-6dad35731caf)

**Comment by Wumpf** [comment]
> I could also imagine people wanting to execute pixi commands using this thing (I don't know what it's called 😅)

Not having a lot of experience either, but I believe this would be a Task Provider:
https://code.visualstudio.com/api/extension-guides/task-provider

(and yes this would be great to have :))

**Comment by pavelzw** [comment]
Another idea could be to automatically use the the recipe-format schema from https://github.com/prefix-dev/recipe-format/blob/main/schema.json for `recipe.yaml` files.

**Reply by pavelzw** [reply]
And ideally also for pixi.toml

**Reply by ruben-arts** [reply]
The `pixi.toml` for sure, but I think the recipe is for now still seperate from `pixi` and in `rattler-build`

**Comment by pavelzw** [comment]
Another thing I would like to see is when you hover over a dependency (or their version) in `pixi.toml`, the latest version of the package and the summary of the package is shown.

![image](https://github.com/prefix-dev/pixi/assets/29506042/e03b29f8-0cde-440b-98d5-41f86c9185d6)

The sparse repodata that bas was talking about could come in handy for this :)

**Comment by lautjy** [comment]
TLDR; Pixi + VSCode results in a jam where VSCode opens hundreds of `rg` processes - needs to be force killed on Macos M1 processor

---

I'll share my user experience in here:

Typically I use VSCode and the inbuilt interactive Python REPL (via ipython kernel), from a `conda env create -f condaenv.yaml`. Interpreter selected from VSCode bottom right menu.

Wanted to try pixi.

* Did `pixi init .` in a folder with 1 .py file. 20 lines of very basic code.
* Then `pixi add ipykernel torchvision pytorch imagehash` .
* `.pixi` folder and `pixi.lock` file appear as expected.

What happens next:
My VSCode gets totally stuck, and keeps opening ripgreps in the background; hundreds of `rg` processes.
Only way to get rid of the situation is Force killing "Code" app and `pkill -f ripgrep` .

Reproduced after rebooting machine and selecting .pixi folder as interpreter.

---

Would there be some way to figure out what goes on?
I'll check if some extension does this, later.

**Reply by pavelzw** [reply]
Shouldn't vscode ignore this directory since it's in the .gitignore? Maybe it doesn't if your directory is not a git project...

**Reply by ruben-arts** [reply]
Is there a way to find out what is trying to be ripgrep'd? And Could you try with only `python` or `ipykernal` for instance?

**Comment by baszalmstra** [comment]
I started integration of pixi into the vscode python plugin: https://github.com/microsoft/vscode-python/pull/22968

Lets see what happens.

**Reply by datajoely** [reply]
Awesome!

**Comment by pavelzw** [comment]
I noticed that the native testing setup doesn't work with vscode since they execute `conda run -p ...` under the hood for conda envs. `conda run -p` currently doesn't work because `conda-meta/history` is missing. If you run `touch .pixi/envs/default/conda-meta/history`, it should work and testing in vscode should work (assuming you have conda installed as well). https://github.com/prefix-dev/pixi/issues/993

**Reply by baszalmstra** [reply]
I am aware of this. My plan is to get the first pr merged in vscode and then add additional features like native module installation (which should better cover the pytest use case).

**Reply by baszalmstra** [reply]
Although Im also not opposed to adding an empty history file.

**Reply by pavelzw** [reply]
i think adding an empty history file would be nice independently of the vscode issue to ensure that `conda run -p` works. maybe even add an end-to-end test?

**Reply by pavelzw** [reply]
https://github.com/prefix-dev/pixi/pull/995

**Comment by traversaro** [comment]
Just to report something else that it would be cool to have on VS Code, that I guess it is not covered by the (awesome) work on VS Code Python extension.

In the team were I work on, we use conda-family tools for C++ development. When working with those in VS Code, the current practice for C++ devs is to open a terminal, activate the conda/pixi environment there, open a `code` session there, and then invoke all the CMake tooling there. This kind of works, but it is not really intuitive and tend to be brittle (as vscode tends to keep some kind of state of the sessions, so you are not always 100% that the right environment is active even if you invoke `code` from a terminal with the active environment. 

It would be cool to have some kind of support that does not require manually invoking `code` from the right terminal. We are exploring some alternatives in https://github.com/ami-iit/ami-commons/issues/22, but we still did not find a reliable solution. I recently discovered `direnv` VS extension, and that seems to work fine together with pixi+direnv on Linux (https://pixi.sh/latest/features/environment/#using-pixi-with-direnv), but it has some rough edges on Windows.

Kind of related to https://github.com/prefix-dev/pixi/issues/411#issuecomment-1837234946 .

**Reply by sdebionne** [reply]
I second that. That would be especially useful if it works with the remote dev extension (where [the only alternative option](https://github.com/microsoft/vscode-cmake-tools/issues/1445) to activate an env using `environmentSetupScript` is quite cumbersome)!

**Reply by traversaro** [reply]
Related twitter comment as well: https://x.com/alvgaona/status/1808878541859074462?t=DvZITo_ZSL9D22HhDNJywQ&s=19 . I guess it may be a good idea to open a dedicated issue or discussion for VSCode + Pixi + C++ support for visibility, as it is a use case quite different from VSCode + Pixi + Python, what do you think prefix.dev people @baszalmstra @wolfv @ruben-arts ?

**Reply by wolfv** [reply]
Yes, that would be great.

**Reply by wolfv** [reply]
Do you want to open that issue? I have looked a little bit into the CMake extension for VSCode to see if we could hook into that ...

**Reply by traversaro** [reply]
Done: https://github.com/prefix-dev/pixi/issues/1570 .

**Reply by traversaro** [reply]
@sdebionne @jfchenier feel free to add a 👍🏽 reaction in https://github.com/prefix-dev/pixi/issues/1570 if you like the feature (so it is easy to track it in https://github.com/prefix-dev/pixi/issues/1281).

**Comment by ruben-arts** [comment]
The vscode python extension is released with pixi support: https://github.com/microsoft/vscode-python/releases/tag/v2024.10.0

**Reply by deltamarnix** [reply]
This is a nice start activating python. I also install other tools via conda, like julia and quarto. These are executables that my visual studio plugins environment cannot find, even when the environment is activated in the terminal. For now I have a .bat file that does `pixi run code .` on my project and then I get everything to work. Would a pixi vs code extension resolve this issue? Or should this already have been resolved by the python plugin?

**Reply by ruben-arts** [reply]
If these tools are handled by their respective plugins this would have to be updated in their plugin. I't curious non of these interpreters are found as they will be just available on the path. 

Could you share your bat file to help others?

**Reply by deltamarnix** [reply]
```bat
pixi run --environment=<your_env> code . | exit
```

I put the `| exit` at the end to ensure that cmd closes after double-clicking. Also, watch the `.` character, that opens the current folder. You could also change this to a `.code-workspace` file if needed.

When I would open vs code from my start menu, I would get errors from the Julia plugin not finding Julia, and from the Quarto plugin not finding Quarto. This is caused by the fact that their default is to look in the PATH, and there is nothing there, since the pixi environment was otherwise not activated. So that means that the pixi environment is activated too late, or in any case, only in the terminals, where they work fine.

vs code still doesn't seem to allow running a startup script for specific projects. Otherwise I could have added that information in the settings.json to call `pixi shell -e <your_env>` just before startup.

**Reply by ruben-arts** [reply]
Ah yeah for the start from start menu we would need an actual pixi extension and not only the python one.
